File: backend/internal/model/tls_fingerprint_router.go
// Package model 定义服务层使用的数据模型。
package model

import (
	"regexp"
	"strconv"
	"strings"
	"time"
)

const (
	// TLSRouterMatchContains 表示 User-Agent 包含 pattern 即命中。
	TLSRouterMatchContains = "contains"
	// TLSRouterMatchPrefix 表示 User-Agent 以 pattern 开头即命中。
	TLSRouterMatchPrefix = "prefix"
	// TLSRouterMatchExact 表示 User-Agent 与 pattern 完全相等才命中。
	TLSRouterMatchExact = "exact"
	// TLSRouterMatchRegex 表示使用正则匹配 User-Agent。
	TLSRouterMatchRegex = "regex"
)

// TLSFingerprintRouterRule 表示 TLS 路由器中的一条 UA 匹配规则。
type TLSFingerprintRouterRule struct {
	Name                    string `json:"name"`
	Enabled                 bool   `json:"enabled"`
	MatchType               string `json:"match_type"`
	Pattern                 string `json:"pattern"`
	CaseSensitive           bool   `json:"case_sensitive"`
	TLSFingerprintProfileID int64  `json:"tls_fingerprint_profile_id"`
	UpstreamUserAgent       string `json:"upstream_user_agent,omitempty"`
	UpstreamOriginator      string `json:"upstream_originator,omitempty"`
}

// TLSFingerprintRouter TLS 指纹路由器。
type TLSFingerprintRouter struct {
	ID                                       int64                      `json:"id"`
	Name                                     string                     `json:"name"`
	Description                              *string                    `json:"description"`
	Enabled                                  bool                       `json:"enabled"`
	ChatGPTOAuthTokenUserAgent               string                     `json:"chatgpt_oauth_token_user_agent"`
	ChatGPTOAuthTokenTLSFingerprintProfileID *int64                     `json:"chatgpt_oauth_token_tls_fingerprint_profile_id"`
	CodexInviteResetUserAgent                string                     `json:"codex_invite_reset_user_agent"`
	CodexInviteResetTLSFingerprintProfileID  *int64                     `json:"codex_invite_reset_tls_fingerprint_profile_id"`
	Rules                                    []TLSFingerprintRouterRule `json:"rules"`
	CreatedAt                                time.Time                  `json:"created_at"`
	UpdatedAt                                time.Time                  `json:"updated_at"`
}

// Validate 验证路由器配置的有效性。
func (r *TLSFingerprintRouter) Validate() error {
	if r.Name == "" {
		return &ValidationError{Field: "name", Message: "name is required"}
	}
	if len(r.ChatGPTOAuthTokenUserAgent) > 512 {
		return &ValidationError{Field: "chatgpt_oauth_token_user_agent", Message: "chatgpt_oauth_token_user_agent must be at most 512 characters"}
	}
	if r.ChatGPTOAuthTokenTLSFingerprintProfileID != nil && *r.ChatGPTOAuthTokenTLSFingerprintProfileID < -1 {
		return &ValidationError{Field: "chatgpt_oauth_token_tls_fingerprint_profile_id", Message: "chatgpt_oauth_token_tls_fingerprint_profile_id must be null, 0, -1, or a positive profile ID"}
	}
	if len(r.CodexInviteResetUserAgent) > 512 {
		return &ValidationError{Field: "codex_invite_reset_user_agent", Message: "codex_invite_reset_user_agent must be at most 512 characters"}
	}
	if r.CodexInviteResetTLSFingerprintProfileID != nil && *r.CodexInviteResetTLSFingerprintProfileID < -1 {
		return &ValidationError{Field: "codex_invite_reset_tls_fingerprint_profile_id", Message: "codex_invite_reset_tls_fingerprint_profile_id must be null, 0, -1, or a positive profile ID"}
	}
	for i, rule := range r.Rules {
		if err := rule.Validate(); err != nil {
			return &ValidationError{Field: "rules", Message: "rule " + strconv.Itoa(i+1) + " " + err.Error()}
		}
	}
	return nil
}

// Validate 验证路由规则配置的有效性。
func (r TLSFingerprintRouterRule) Validate() error {
	if strings.TrimSpace(r.Pattern) == "" {
		return &ValidationError{Field: "pattern", Message: "pattern is required"}
	}
	switch NormalizeTLSRouterMatchType(r.MatchType) {
	case TLSRouterMatchContains, TLSRouterMatchPrefix, TLSRouterMatchExact:
		return nil
	case TLSRouterMatchRegex:
		if _, err := regexp.Compile(r.Pattern); err != nil {
			return &ValidationError{Field: "pattern", Message: "regex is invalid"}
		}
		return nil
	default:
		return &ValidationError{Field: "match_type", Message: "match_type must be contains, prefix, exact, or regex"}
	}
}

// NormalizeTLSRouterMatchType 归一化 TLS 路由器匹配类型。
func NormalizeTLSRouterMatchType(matchType string) string {
	switch strings.ToLower(strings.TrimSpace(matchType)) {
	case TLSRouterMatchPrefix:
		return TLSRouterMatchPrefix
	case TLSRouterMatchExact:
		return TLSRouterMatchExact
	case TLSRouterMatchRegex:
		return TLSRouterMatchRegex
	default:
		return TLSRouterMatchContains
	}
}

File: frontend/src/api/admin/tlsFingerprintRouter.ts
/**
 * 管理端 TLS 指纹路由器 API。
 * 用于维护基于 User-Agent 匹配的 TLS 指纹路由器。
 */

import { apiClient } from '../client'

export type TLSFingerprintRouterMatchType = 'contains' | 'prefix' | 'exact' | 'regex'

export interface TLSFingerprintRouterRule {
  name: string
  enabled: boolean
  match_type: TLSFingerprintRouterMatchType
  pattern: string
  case_sensitive: boolean
  tls_fingerprint_profile_id: number
  upstream_user_agent?: string
  upstream_originator?: string
}

export interface TLSFingerprintRouter {
  id: number
  name: string
  description: string | null
  enabled: boolean
  chatgpt_oauth_token_user_agent?: string
  chatgpt_oauth_token_tls_fingerprint_profile_id?: number | null
  codex_invite_reset_user_agent?: string
  codex_invite_reset_tls_fingerprint_profile_id?: number | null
  rules: TLSFingerprintRouterRule[]
  created_at: string
  updated_at: string
}

export interface CreateTLSFingerprintRouterRequest {
  name: string
  description?: string | null
  enabled?: boolean
  chatgpt_oauth_token_user_agent?: string
  chatgpt_oauth_token_tls_fingerprint_profile_id?: number | null
  codex_invite_reset_user_agent?: string
  codex_invite_reset_tls_fingerprint_profile_id?: number | null
  rules?: TLSFingerprintRouterRule[]
}

export interface UpdateTLSFingerprintRouterRequest {
  name?: string
  description?: string | null
  enabled?: boolean
  chatgpt_oauth_token_user_agent?: string
  chatgpt_oauth_token_tls_fingerprint_profile_id?: number | null
  codex_invite_reset_user_agent?: string
  codex_invite_reset_tls_fingerprint_profile_id?: number | null
  rules?: TLSFingerprintRouterRule[]
}

export async function list(): Promise<TLSFingerprintRouter[]> {
  const { data } = await apiClient.get<TLSFingerprintRouter[]>('/admin/tls-fingerprint-routers')
  return data
}

export async function getById(id: number): Promise<TLSFingerprintRouter> {
  const { data } = await apiClient.get<TLSFingerprintRouter>(`/admin/tls-fingerprint-routers/${id}`)
  return data
}

export async function create(payload: CreateTLSFingerprintRouterRequest): Promise<TLSFingerprintRouter> {
  const { data } = await apiClient.post<TLSFingerprintRouter>('/admin/tls-fingerprint-routers', payload)
  return data
}

export async function update(id: number, payload: UpdateTLSFingerprintRouterRequest): Promise<TLSFingerprintRouter> {
  const { data } = await apiClient.put<TLSFingerprintRouter>(`/admin/tls-fingerprint-routers/${id}`, payload)
  return data
}

export async function deleteRouter(id: number): Promise<{ message: string }> {
  const { data } = await apiClient.delete<{ message: string }>(`/admin/tls-fingerprint-routers/${id}`)
  return data
}

export const tlsFingerprintRouterAPI = {
  list,
  getById,
  create,
  update,
  delete: deleteRouter
}

export default tlsFingerprintRouterAPI
